File: 백준/Gold/3109. 빵집/빵집.py
import sys
input = sys.stdin.readline


# 오른쪽, 오른쪽 위대각선, 오른쪽 아래 대각선
# [0, 1] [-1, 1], [1, 1]

def dfs(x, y):
    
    visited[x][y] = True

    if y == c - 1:
        return True

    dx, dy = [-1, 0, 1], [1, 1, 1]
    for i in range(3):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0 <= nx < r and pipeline[nx][ny] == '.' and not visited[nx][ny]:
            visited[nx][ny] = True
            if dfs(nx, ny):
                return True

    return False


r, c = map(int, input().split())
pipeline = [list(input()) for _ in range(r)]

cnt = 0
visited = [[False] * c for _ in range(r)]
for a in range(r):
    if pipeline[a][0] == '.':
        if dfs(a, 0):
            cnt += 1

print(cnt)

File: 백준/Gold/3109. 빵집/빵집.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {

	private static String[][] map;
	private static boolean[][] visited;
	private static int r, c;

	public static void main(String[] args) throws Exception {

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		String[] split = in.readLine().split(" ");

		r = Integer.parseInt(split[0]);
		c = Integer.parseInt(split[1]);

		map = new String[r][c];
		for (int i = 0; i < r; i++) {
			split = in.readLine().split("");
			for (int j = 0; j < c; j++) {
				map[i][j] = split[j];
			}
		}

		visited = new boolean[r][c];
		int cnt = 0;
		for (int i = 0; i < r; i++) {
			if (dfs(i, 0)) {
				cnt++;
//				for (boolean[] v : visited) {
//					System.out.println(Arrays.toString(v));
//				}
			}

		}

		System.out.println(cnt);

	}

	private static int[] dx = { -1, 0, 1 }; // 우상, 오, 우하
	private static int[] dy = { 1, 1, 1 };

	private static boolean dfs(int x, int y) {

		visited[x][y] = true;
		if (y == c - 1) {
			return true;
		}

		for (int i = 0; i < 3; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (nx < 0 || nx >= r || ny < 0 || ny >= c) {
				continue;
			}

			if (!visited[nx][ny] && map[nx][ny].equals(".")) {
				visited[nx][ny] = true;
				if (dfs(nx, ny)) {
					return true;
				}
			}
		}

		return false;
	}

}